python_modules/dagster/dagster/_core/asset_graph_view/asset_graph_view.py

```diff
@@ -0,0 +1,328 @@
+from datetime import datetime
+from typing import TYPE_CHECKING, AbstractSet, Iterable, Mapping, NamedTuple, NewType, Optional
+
+from dagster import _check as check
+from dagster._core.definitions.asset_subset import AssetSubset, ValidAssetSubset
+from dagster._core.definitions.events import AssetKey, AssetKeyPartitionKey
+from dagster._utils.cached_method import cached_method
+
+if TYPE_CHECKING:
+    from dagster._core.definitions.base_asset_graph import BaseAssetGraph, BaseAssetNode
+    from dagster._core.definitions.data_version import CachingStaleStatusResolver
+    from dagster._core.definitions.definitions_class import Definitions
+    from dagster._core.definitions.partition import PartitionsDefinition
+    from dagster._core.instance import DagsterInstance
+    from dagster._utils.caching_instance_queryer import CachingInstanceQueryer
+
+
+class TemporalContext(NamedTuple):
+    """TemporalContext represents an effective time, used for business logic, and last_event_id
+    which is used to identify that state of the event log at some point in time. Put another way,
+    the value of a TemporalContext represents a point in time and a snapshot of the event log.
+
+    Effective time: This is the effective time of the computation in terms of business logic,
+    and it impacts the behavior of partioning and partition mapping. For example, if an
+    if you get the "last" partition window of a given partitions definition, it is with
+    respect to the effective time.
+
+    Last event id: Our event log has a monotonic increasing event id. This is used to
+    cursor the event log. This event_id is also propogated to derived tables to indicate.
+    This allows us to query the state of the event log at a given point in time.
+
+    Note that insertion time of the last_event_id is not the same as the effective time.
+
+    A last_event_id of None indicates that the reads will be volatile will immediately
+    reflect any subsequent writes.
+    """
+
+    effective_dt: datetime
+    last_event_id: Optional[int]
+
+
+class _AssetSliceCompatibleSubset(ValidAssetSubset): ...
+
+
+PartitionKey = NewType("PartitionKey", str)
+"""
+PartitionKey is a string that represents a partition key. It is a new type so
+that we can track the use of strs as partition keys.
+
+PartitionKey should only be created via the static factory methods on AssetPartitionKey.
+We want to consolidate all partition key creation so we have a leverage point to change
+how they. Eventually we want to create a new singleton string partition to represent
+the "all" partition key for unpartitioned assets.
+"""
+
+
+class AssetPartitionKey(NamedTuple):
+    """Represents a a partition in a particular asset. Prefer usage to AssetKeyPartitionKey.
+    If you need to convert AssetKeyPartitionKeys to AssetPartitionKeys, use the static
+    factory methods.
```

Comment:
@OwenKephart and @smackesey would like your sign-off here. I want to make sure that we are aligned that this executes on the plan we discussed in the Slack huddle.

Reply:
(The docblock on `AssetPartitionKey` and `PartitionKey`